src/Languages/Core/Impl/Formatting/Whitespace.cs

```diff
@@ -110,5 +111,28 @@ public static class Whitespace {
 
             return false;
         }
+
+        public static string NormalizeWhitespace(this string s) {
+            if(s == null || s.Length == 0) {
+                return s;
+            }
+
+            var cs = new CharacterStream(new TextStream(s));
+            var sb = new StringBuilder();
+
+            while (!cs.IsEndOfStream()) {
+                var current = cs.Position;
+                cs.SkipWhitespace();
+                if (cs.Position - current > 0) {
+                    sb.Append(' ');
+                }
+
+                while (!cs.IsEndOfStream() && !cs.IsWhiteSpace()) {
```

Comment:
There can be one loop instead of two, but I'm not sure that it will be easier to read, so up to you